src/benchmark/file_io_read_micro_benchmark.cpp

```diff
@@ -433,120 +400,122 @@ void FileIOMicroReadBenchmarkFixture::aio_single_threaded(benchmark::State& stat
   close(fd);
 }
 
-void FileIOMicroReadBenchmarkFixture::aio_multi_threaded(benchmark::State& state, uint16_t thread_count) {
-  auto fd = int32_t{};
-  Assert(((fd = open(filename, O_RDONLY)) >= 0), close_file_and_return_error_message(fd, "Open error: ", errno));
+void FileIOMicroReadBenchmarkFixture::aio_sequential_read_multi_threaded(benchmark::State& state, uint16_t aio_request_count) {
+  auto filedescriptors = std::vector<int32_t>(aio_request_count);
+  for (auto index = size_t{0}; index < aio_request_count; ++index) {
+    auto fd = int32_t{};
+    Assert(((fd = open(filename, O_RDONLY)) >= 0), close_file_and_return_error_message(fd, "Open error: ", errno));
+    filedescriptors[index] = fd;
+  }
 
-  auto threads = std::vector<std::thread>(thread_count);
-  auto batch_size = static_cast<uint64_t>(std::ceil(static_cast<float>(NUMBER_OF_ELEMENTS) / thread_count));
+  auto batch_size = static_cast<uint64_t>(std::ceil(static_cast<float>(NUMBER_OF_ELEMENTS) / aio_request_count));
+
+  // init aio to only use specified amounts of threads (not part of POSIX API, only defined in GNU-C libary)
+#ifdef __linux__
+  static struct aioinit init_data;
+    init_data.aio_threads = aio_request_count;
+    aio_init(&init_data);
+#endif
 
   for (auto _ : state) {
     state.PauseTiming();
-
-    micro_benchmark_clear_disk_cache();
     auto read_data = std::vector<uint32_t>{};
     read_data.resize(NUMBER_OF_ELEMENTS);
-    auto* read_data_start = std::data(read_data);
-
+    micro_benchmark_clear_disk_cache();
     state.ResumeTiming();
 
-    for (auto index = size_t{0}; index < thread_count; ++index) {
+    auto aio = std::vector<aiocb>(aio_request_count);
+    auto aio_list = std::vector<aiocb*>(aio_request_count);
+
+    auto read_data_ptr = std::data(read_data);
+
+    for (auto index = size_t{0}; index < aio_request_count; ++index) {
       auto from = batch_size * index;
       auto to = from + batch_size;
       if (to >= NUMBER_OF_ELEMENTS) {
         to = NUMBER_OF_ELEMENTS;
       }
-      threads[index] = (std::thread(read_data_using_aio, from, to, fd, read_data_start));
-    }
-
-    for (auto index = size_t{0}; index < thread_count; ++index) {
-      // Explain: Blocks the current thread until the thread identified by *this finishes its execution
-      threads[index].join();
+      create_aio_request(aio[index], filedescriptors[index], from * uint32_t_size, read_data_ptr + from, (to - from) * uint32_t_size, LIO_READ);
```

Comment:
Just to get this right: `create_aio_request` just "registers a I/O request`, but doesn't execute it? This is done via `lio_listio`?

Reply:
Yes, exactly. lio_listio initiates the sending of all requests and waits for them.